### ISSUE TITLE
assign_op_pattern misleading on wrongly implemented assign impl.

### ISSUE THREAD

**Issue by letheed** [rust-lang/rust-clippy#2268]
```rust
impl SubAssign for Foo {
    fn sub_assign(&mut self, rhs: Self) {
        *self = *self + rhs; // Copy-paste facepalm
    }
}
```

Triggers assign_op_pattern, which suggest `replace it with: *self += rhs`.

Mistake caught but left me wondering for a minute. Had to find [this](https://github.com/rust-lang-nursery/rust-clippy/pull/1318) to figure out something else must be going on.

**Comment by oli-obk** [comment]
duplicate of #1239

**Comment by oli-obk** [comment]
Oh, this is `SubAssign` not `AddAssign`, nevermind me

**Comment by Ryan1729** [comment]
Given that #2458, which addresses this, landed some time ago, and that #5820 which reduces the false positive rate on the `suspicious_arithmetic_impl` lint, recently landed, can this issue be closed?

#2458 does mention that the `suspicious_arithmetic_impl` lint only covers the `Add`, `Sub`, `Mul` and `Div`, traits, and [this is still true currently](https://github.com/ThibsG/rust-clippy/blob/79f948ec0a507f8dc663508ce013104847fcc9f3/clippy_lints/src/suspicious_trait_impl.rs#L93-L98). Do we want to extend this lint to cover the rest of the binary operations that can be overloaded? Specifically those are `BitAnd`, `BitOr`, `BitXor`, `Rem`, `Shl`, and `Shr`.

**Comment by andrewjstone** [comment]
For any one performing these operations on types are not over the integers, like in finite fields, this lint gives false positives as well. 

`^` is valid in `Add` and `Sub` and `*` is valid in `Div`. This feels like a lint that's very hard to get right. Importantly it inflicts unnecessary pain on people doing the correct thing in order to avoid copy paste bugs, which feels backwards to me.
